### ISSUE TITLE
goANI -> KeyError

### ISSUE THREAD

**Issue by n1mus** [MrOlm/drep#68]
Hi,

I was testing dRep and I noticed that using goANI throws an error?

`dRep dereplicate /kb/module/work/tmp/dRep_workDir_59fa9163-bebd-47de-887e-77bee7c112f3 --genomes /kb/module/work/tmp/binsPooled_59fa9163-bebd-47de-887e-77bee7c112f3/* --debug --processors 16 --S_algorithm goANI`

gives

```
***************************************************
    ..:: dRep dereplicate Step 1. Filter ::..
***************************************************
    
Will filter the genome list
Calculating genome info of genomes
100.00% of genomes passed length filtering
Running prodigal
Running checkM
100.00% of genomes passed checkM filtering
***************************************************
    ..:: dRep dereplicate Step 2. Cluster ::..
***************************************************
    
Clustering Step 1. Parse Arguments
Clustering Step 2. Perform MASH (primary) clustering
2a. Run pair-wise MASH clustering
2b. Cluster pair-wise MASH clustering
8 primary clusters made
Step 3. Perform secondary clustering
Running 17 goANI comparisons- should take ~ 0.1 min
Past prodigal runs found- will not re-run
Traceback (most recent call last):
  File "/miniconda/bin/dRep", line 33, in <module>
    controller.parseArguments(args)
  File "/miniconda/lib/python3.6/site-packages/drep/controller.py", line 144, in parseArguments
    self.dereplicate_operation(**vars(args))
  File "/miniconda/lib/python3.6/site-packages/drep/controller.py", line 86, in dereplicate_operation
    drep.d_workflows.dereplicate_wrapper(kwargs['work_directory'],**kwargs)
  File "/miniconda/lib/python3.6/site-packages/drep/d_workflows.py", line 36, in dereplicate_wrapper
    drep.d_cluster.d_cluster_wrapper(wd, **kwargs)
  File "/miniconda/lib/python3.6/site-packages/drep/d_cluster.py", line 78, in d_cluster_wrapper
    data_folder, wd=workDirectory, **kwargs)
  File "/miniconda/lib/python3.6/site-packages/drep/d_cluster.py", line 200, in cluster_genomes
    ndb = compare_genomes(bdb, algorithm, data_folder, **kwargs)
  File "/miniconda/lib/python3.6/site-packages/drep/d_cluster.py", line 918, in compare_genomes
    prod_folder=prod_folder, **kwargs)
  File "/miniconda/lib/python3.6/site-packages/drep/d_cluster.py", line 1208, in run_pairwise_goANI
    df = process_goani_files(files)
  File "/miniconda/lib/python3.6/site-packages/drep/d_cluster.py", line 785, in process_goani_files
    results = parse_nsim_file(file)
  File "/miniconda/lib/python3.6/site-packages/drep/d_cluster.py", line 1267, in parse_nsim_file
    db = _filter_nsimscan(db)
  File "/miniconda/lib/python3.6/site-packages/drep/d_cluster.py", line 1278, in _filter_nsimscan
    db1 = db1.sort_values(['al_len','qry_id', 'sbj_id'], ascending=False)
  File "/miniconda/lib/python3.6/site-packages/pandas/core/frame.py", line 5001, in sort_values
    keys = [self._get_label_or_level_values(x, axis=axis) for x in by]
  File "/miniconda/lib/python3.6/site-packages/pandas/core/frame.py", line 5001, in <listcomp>
    keys = [self._get_label_or_level_values(x, axis=axis) for x in by]
  File "/miniconda/lib/python3.6/site-packages/pandas/core/generic.py", line 1774, in _get_label_or_level_values
    raise KeyError(key)
KeyError: 'qry_id'
```

Am I doing this right?

Best wishes,
Sumin

**Comment by MrOlm** [comment]
Hello,

Thanks for pointing out- this has been fixed in version 2.4.1. Please let me know if you're still having a problem.

Best,
Matt

**Comment by n1mus** [comment]
Hi,

I pip installed 2.4.1 but I'm still getting the same behavior for `--S_algorithm goANI`. [My test file](https://github.com/n1mus/dRep/blob/4fd7587e772e8e42b1f229fbb4d3b5ab0ec75454/test/dRep_server_test.py).

Upon further testing, `--run_tax` is also crashing, i.e.,

`dRep dereplicate /kb/module/work/tmp/dRep_workDir_b85556ca-0898-4e5e-9d2f-398dbb80e5f1 --genomes /kb/module/work/tmp/binsPooled_b85556ca-0898-4e5e-9d2f-398dbb80e5f1/*.fasta --debug --processors 16 --run_tax --tax_method max --percent 55`

gives

```
***************************************************
    ..:: dRep dereplicate Step 1. Filter ::..
***************************************************

Will filter the genome list
Calculating genome info of genomes
100.00% of genomes passed length filtering
Running prodigal
Running checkM
56.47% of genomes passed checkM filtering
***************************************************
    ..:: dRep dereplicate Step 2. Cluster ::..
***************************************************

Clustering Step 1. Parse Arguments
Clustering Step 2. Perform MASH (primary) clustering
2a. Run pair-wise MASH clustering
2b. Cluster pair-wise MASH clustering
26 primary clusters made
Step 3. Perform secondary clustering
Running 92 ANImf comparisons- should take ~ 2.9 min
Step 4. Return output
***************************************************
    ..:: dRep dereplicate Step 3. Choose ::..
***************************************************

Loading work directory
***************************************************
    ..:: dRep dereplicate Step 4. Bonus ::..
***************************************************

Loading work directory
Running tax
Past prodigal runs found- will not re-run
Running Centrifuge
Traceback (most recent call last):
  File "/miniconda/bin/dRep", line 33, in <module>
    controller.parseArguments(args)
  File "/miniconda/lib/python3.6/site-packages/drep/controller.py", line 144, in parseArguments
    self.dereplicate_operation(**vars(args))
  File "/miniconda/lib/python3.6/site-packages/drep/controller.py", line 86, in dereplicate_operation
    drep.d_workflows.dereplicate_wrapper(kwargs['work_directory'],**kwargs)
  File "/miniconda/lib/python3.6/site-packages/drep/d_workflows.py", line 52, in dereplicate_wrapper
    drep.d_bonus.d_bonus_wrapper(wd, **kwargs)
  File "/miniconda/lib/python3.6/site-packages/drep/d_bonus.py", line 26, in d_bonus_wrapper
    run_taxonomy(wd,**kwargs)
  File "/miniconda/lib/python3.6/site-packages/drep/d_bonus.py", line 36, in run_taxonomy
    run_centrifuge(Bdb, prod_dir, cent_dir, wd=wd, **kwargs)
  File "/miniconda/lib/python3.6/site-packages/drep/d_bonus.py", line 139, in run_centrifuge
    logging.debug(' '.join(cmd))
TypeError: sequence item 3: expected str instance, NoneType found
```

and

`Rep dereplicate /kb/module/work/tmp/dRep_workDir_dd30dbfb-f7cc-4860-a45c-3a3c1668c7d8 --genomes /kb/module/work/tmp/binsPooled_dd30dbfb-f7cc-4860-a45c-3a3c1668c7d8/*.fasta --debug --processors 16 --run_tax --percent 55`

gives

```
***************************************************
    ..:: dRep dereplicate Step 1. Filter ::..
***************************************************

Will filter the genome list
Calculating genome info of genomes
100.00% of genomes passed length filtering
Running prodigal
Running checkM
56.47% of genomes passed checkM filtering
***************************************************
    ..:: dRep dereplicate Step 2. Cluster ::..
***************************************************

Clustering Step 1. Parse Arguments
Clustering Step 2. Perform MASH (primary) clustering
2a. Run pair-wise MASH clustering
2b. Cluster pair-wise MASH clustering
26 primary clusters made
Step 3. Perform secondary clustering
Running 92 ANImf comparisons- should take ~ 2.9 min
Step 4. Return output
***************************************************
    ..:: dRep dereplicate Step 3. Choose ::..
***************************************************

Loading work directory
***************************************************
    ..:: dRep dereplicate Step 4. Bonus ::..
***************************************************

Loading work directory
Running tax
Past prodigal runs found- will not re-run
Running Centrifuge
Traceback (most recent call last):
  File "/miniconda/bin/dRep", line 33, in <module>
    controller.parseArguments(args)
  File "/miniconda/lib/python3.6/site-packages/drep/controller.py", line 144, in parseArguments
    self.dereplicate_operation(**vars(args))
  File "/miniconda/lib/python3.6/site-packages/drep/controller.py", line 86, in dereplicate_operation
    drep.d_workflows.dereplicate_wrapper(kwargs['work_directory'],**kwargs)
  File "/miniconda/lib/python3.6/site-packages/drep/d_workflows.py", line 52, in dereplicate_wrapper
    drep.d_bonus.d_bonus_wrapper(wd, **kwargs)
  File "/miniconda/lib/python3.6/site-packages/drep/d_bonus.py", line 26, in d_bonus_wrapper
    run_taxonomy(wd,**kwargs)
  File "/miniconda/lib/python3.6/site-packages/drep/d_bonus.py", line 36, in run_taxonomy
    run_centrifuge(Bdb, prod_dir, cent_dir, wd=wd, **kwargs)
  File "/miniconda/lib/python3.6/site-packages/drep/d_bonus.py", line 139, in run_centrifuge
    logging.debug(' '.join(cmd))
TypeError: sequence item 3: expected str instance, NoneType found
```

My installations are all there with the `dRep bonus testDir --check_dependencies` and visual confirmation, and I'm running this in a docker. I hope it's not just my dataset. But at least everything else seems to be working very well.


Best wishes,
Sumin

**Comment by MrOlm** [comment]
Hi Sumin,

1) Could you please send me the error that goANI is giving you? Similar to the tracebacks that you provided for centrifuge failures?

2) To run taxonomy, you also need to provide a centrifuge index using the flag "--cent_index". Though to be honest the taxonomy values that it gives aren't very accurate, so you might want to consider a different way of figuring out taxonomy.

Thank you,
-Matt

**Comment by n1mus** [comment]
Hi Matt,

It looks like the same behavior, but with a different dataset this time (capybara gut microbiome binned through MaxBin2 and MetaBAT2) The command `dRep dereplicate /kb/module/work/tmp/dRep_workDir_8d6dfd3f-3807-4786-857f-5767ac9596b7 -
-genomes /kb/module/work/tmp/binsPooled_8d6dfd3f-3807-4786-857f-5767ac9596b7/*.fasta --debug --processors 16 --S_algorithm goA
NI` yielded

```
***************************************************
    ..:: dRep dereplicate Step 1. Filter ::..
***************************************************

Will filter the genome list
Calculating genome info of genomes
100.00% of genomes passed length filtering
Running prodigal
Running checkM
46.75% of genomes passed checkM filtering
***************************************************
    ..:: dRep dereplicate Step 2. Cluster ::..
***************************************************

Clustering Step 1. Parse Arguments
Clustering Step 2. Perform MASH (primary) clustering
2a. Run pair-wise MASH clustering
2b. Cluster pair-wise MASH clustering
49 primary clusters made
Step 3. Perform secondary clustering
Running 139 goANI comparisons- should take ~ 0.9 min
Past prodigal runs found- will not re-run
Past prodigal runs found- will not re-run
Past prodigal runs found- will not re-run
Past prodigal runs found- will not re-run
Traceback (most recent call last):
  File "/miniconda/bin/dRep", line 33, in <module>
    controller.parseArguments(args)
  File "/miniconda/lib/python3.6/site-packages/drep/controller.py", line 144, in parseArguments
    self.dereplicate_operation(**vars(args))
  File "/miniconda/lib/python3.6/site-packages/drep/controller.py", line 86, in dereplicate_operation
    drep.d_workflows.dereplicate_wrapper(kwargs['work_directory'],**kwargs)
  File "/miniconda/lib/python3.6/site-packages/drep/d_workflows.py", line 36, in dereplicate_wrapper
    drep.d_cluster.d_cluster_wrapper(wd, **kwargs)
  File "/miniconda/lib/python3.6/site-packages/drep/d_cluster.py", line 78, in d_cluster_wrapper
    data_folder, wd=workDirectory, **kwargs)
  File "/miniconda/lib/python3.6/site-packages/drep/d_cluster.py", line 200, in cluster_genomes
    ndb = compare_genomes(bdb, algorithm, data_folder, **kwargs)
  File "/miniconda/lib/python3.6/site-packages/drep/d_cluster.py", line 918, in compare_genomes
    prod_folder=prod_folder, **kwargs)
  File "/miniconda/lib/python3.6/site-packages/drep/d_cluster.py", line 1208, in run_pairwise_goANI
    df = process_goani_files(files)
  File "/miniconda/lib/python3.6/site-packages/drep/d_cluster.py", line 785, in process_goani_files
    results = parse_nsim_file(file)
  File "/miniconda/lib/python3.6/site-packages/drep/d_cluster.py", line 1273, in parse_nsim_file
    db = _filter_nsimscan(db)
  File "/miniconda/lib/python3.6/site-packages/drep/d_cluster.py", line 1284, in _filter_nsimscan
    db1 = db1.sort_values(['al_len','qry_id', 'sbj_id'], ascending=False)
  File "/miniconda/lib/python3.6/site-packages/pandas/core/frame.py", line 5001, in sort_values
    keys = [self._get_label_or_level_values(x, axis=axis) for x in by]
  File "/miniconda/lib/python3.6/site-packages/pandas/core/frame.py", line 5001, in <listcomp>
    keys = [self._get_label_or_level_values(x, axis=axis) for x in by]
  File "/miniconda/lib/python3.6/site-packages/pandas/core/generic.py", line 1774, in _get_label_or_level_values
    raise KeyError(key)
KeyError: 'qry_id'
```
I tried attaching the work directory but it was too big.

Okay thanks for the advice, I am getting input that I should probably just drop centrifuge, because it has tiny reference data and people may not actively desire using it.


Best wishes,
Sumin

**Comment by MrOlm** [comment]
Hi Sumin,

Thanks you for your help- this bug is very difficult to track down. I've just pushed a new branch called `testingBranch` that I'd like you to try- this will give much more information about what the problem is.

To install it please install the branch "testingBranch" of dRep. You can do with the command

`git clone -b testingBranch git@github.com:MrOlm/drep.git`

You can then install it by going into the drep folder and running the command

`pip install . --upgrade`

If all goes well, you should see the following dRep version number when you run the help:

```$ dRep -h

                ...::: dRep v2.4.1_TESTING :::...

  Matt Olm. MIT License. Banfield Lab, UC Berkeley. 2017

  Choose one of the operations below for more detailed help. See https://drep.readthedocs.io/en/latest/index.html for documentation
  Example: dRep dereplicate -h

  Workflows:
    compare         -> Perform rapid pair-wise comparison on a list of genomes
    dereplicate     -> De-replicate a list of genomes

  Single operations:
    filter          -> Filter a genome list based on size, completeness, and/or contamination
    cluster         -> Compare and cluster a genome list based on MASH and ANIn/gANI
    choose          -> Choose the best genome from each genome cluster
    evaluate        -> Evaluate genome de-replication
    bonus           -> Other random operations (determine taxonomy / check dependencies)
    analyze         -> Make figures```

Once installed, please re-run and tell me what the output shows. It should specifically specify where the problem is occurring.

Best,
Matt

**Comment by n1mus** [comment]
Hi Matt,

I had trouble with those commands, but I did manage to `cp` the source files from the testing branch into my local python package and get the correct help page.

So I ran this: `dRep dereplicate dRep_workDir_capybaraGut.2binners --genomes capybaraGut*/*.fasta --debug --processors 8 --checkM_method taxonomy_wf --S_algorithm goANI` and got

```
***************************************************
    ..:: dRep dereplicate Step 1. Filter ::..
***************************************************
    
Will filter the genome list
Calculating genome info of genomes
100.00% of genomes passed length filtering
Running prodigal
Running checkM
44.38% of genomes passed checkM filtering
***************************************************
    ..:: dRep dereplicate Step 2. Cluster ::..
***************************************************
    
Clustering Step 1. Parse Arguments
Clustering Step 2. Perform MASH (primary) clustering
2a. Run pair-wise MASH clustering
2b. Cluster pair-wise MASH clustering
45 primary clusters made
Step 3. Perform secondary clustering
Running 135 goANI comparisons- should take ~ 1.7 min
Past prodigal runs found- will not re-run
Past prodigal runs found- will not re-run
Past prodigal runs found- will not re-run
ERROR! FILE /home/sumin/kbsdk-workspace/dRep/test_local/uncompressed_test_data/dRep_workDir_capybaraGut.2binners/data/goANI_files/Bin.002.fasta/Bin.002.fasta_vs_bin.058.fasta.sim
                    #Q_id                  S_id  p_inden  al_len  mism  gaps  gap_len  qry_beg  qry_end  qry_len  trg_beg  trg_end  trg_len  evalue  sw_score  qry_auto  trg_auto               CIGAR
0     Ga0129306_1000343_1   Ga0129306_1000343_1   100.00    1587     0     0        0        1     1587     1587        1     1587     1587  4800.0    1587.0      1587      1587               1587M
1     Ga0129306_1000343_1   Ga0129306_1000343_2    77.78     252    56     0        0     1336     1587     1587     1327     1578     1578   370.0     140.0       252       252                252M
2     Ga0129306_1000343_1   Ga0129306_1004293_7    67.42     310   101     1        3     1275     1587     1587     1539     1848     1848   290.0     108.0       310       310           40M3I270M
3     Ga0129306_1000343_1  Ga0129306_1000729_41    70.00     270    81     3       10     1239     1513     1587     1464     1735     1806   290.0     108.0       270       270  94M3I75M2I10M2D91M
4     Ga0129306_1000343_1  Ga0129306_1000729_40    69.96     263    79     3       17     1239     1506     1587     1194     1458     1596   280.0     105.0       263       263  71M3I98M2I10M2D84M
...                   ...                   ...      ...     ...   ...   ...      ...      ...      ...      ...      ...      ...      ...     ...       ...       ...       ...                 ...
4148  Ga0129306_1029647_1   Ga0129306_1005920_1   100.00     120     0     0     4002     1255     1374     1374        1      120      435   360.0     120.0       120       120                120M
4149  Ga0129306_1029881_2   Ga0129306_1001285_5    65.59      93    32     0     4002      125      217      723      128      220      759    84.0      29.0        93        93                 93M
4150  Ga0129306_1030529_1   Ga0129306_1008690_1   100.00     120     0     0     4002      430      549      549        1      120      276   360.0     120.0       120       120                120M
4151  Ga0129306_1031959_1   Ga0129306_1008153_5   100.00     120     0     0     4002      598      717      717        1      120     1065   360.0     120.0       120       120                120M
4152  Ga0129306_1031959_2  Ga0129306_1003950_10   100.00     123     0     0     4002        1      123      567      910     1032     1032   370.0     123.0       123       123                123M

[4153 rows x 18 columns]
```

Let me know if you need anything else! I had a bit of a headache duplicating the exact environment and accidentally deleted the work directory, but I'm happy to help.

Best,
Sumin

**Comment by MrOlm** [comment]
Hi Sumin,

Thank you for your help catching this truely bizarre bug. I believe I have fixed the issue in dRep version 2.4.2, which I've just pushed to github and pypi (so you can update by running `pip install drep --upgrade`

Best,
Matt
